src/Query/Select.php

```diff
@@ -0,0 +1,637 @@
+    }
+
+    // HAVING
+    if (count($this->having)) {
+      $query .= " HAVING " . $this->havingString;
+    }
+
+    // ORDER BY
+    if ($this->order) {
+      $query .= " ORDER BY ";
+      $fields = [];
+      foreach ($this->order as $field => $direction) {
+        $fields[] = $field . ' ' . $direction;
+      }
+      $query .= implode(', ', $fields);
+    }
+
+    // RANGE
+    // There is no universal SQL standard for handling range or limit clauses.
+    // Fortunately, all core-supported databases use the same range syntax.
+    // Databases that need a different syntax can override this method and
+    // do whatever alternate logic they need to.
+    if (!empty($this->range)) {
+      $query .= " LIMIT " . (int) $this->range['length'] . " OFFSET " . (int) $this->range['start'];
+    }
+
+    if ($this->forUpdate) {
+      $query .= ' FOR UPDATE';
+    }
+
+    return $query;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __clone() {
+    $this->condition = clone($this->condition);
+    $this->having = clone($this->having);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function &getTables() {
+    return $this->tables;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function &getFields() {
+    return $this->fields;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function &getExpressions() {
+    return $this->expressions;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function &getOrderBy() {
+    return $this->order;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function &getGroupBy() {
+    return $this->group;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function condition($field, $value = NULL, $operator = '=') {
+    $this->condition->condition($field, $value, $operator);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isNull($field) {
+    $this->condition->isNull($field);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isNotNull($field) {
+    $this->condition->isNotNull($field);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function alwaysFalse() {
+    $this->condition->alwaysFalse();
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function &conditions() {
+    return $this->condition->conditions();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function where($snippet) {
+    $this->condition->where($snippet);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function conditionGroupFactory($conjunction = 'AND') {
+    return new Condition($conjunction);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function andConditionGroup() {
+    return $this->conditionGroupFactory('AND');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function orConditionGroup() {
+    return $this->conditionGroupFactory('OR');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function addTable($table, $alias = NULL) {
```

Comment:
Can we actually add tables in this fashion? I think this might have to be handled on the field list

Reply:
Yeah, sort of, e.g. if you want to add a table in order to create a where condition, but not select anything from it.

Here's an example query from [https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_alias.htm?search_text=alias](https://developer.salesforce.com/docs/atlas.en-us.soql_sosl.meta/soql_sosl/sforce_api_calls_soql_alias.htm?search_text=alias)

```
SELECT count()
FROM Contact c, c.Account a
WHERE a.name = 'MyriadPubs'

```